tests/system/marathon_common_tests.py

```diff
@@ -726,10 +731,56 @@ def check_task_recovery():
     port = tasks[0]['ports'][0]
     host = tasks[0]['host']
     cmd = "curl {}:{}/data/foo".format(host, port)
-    run, data = shakedown.run_command_on_master(cmd)
 
-    assert run, "{} did not succeed".format(cmd)
-    assert data == 'hello\nhello\n', "'{}' was not equal to hello\\nhello\\n".format(data)
+    check_task(cmd, target_data='hello\nhello\n')
+
+
+@shakedown.dcos_1_8
+def test_app_with_persistent_volume_recovers():
+    """Tests that when an app task with a persistent volume gets killed,
+       it recovers on the node it was launched on, and it gets attached
+       to the same persistent-volume."""
+
+    app_def = apps.persistent_volume_app()
+    app_id = app_def['id']
+
+    client = marathon.create_client()
+    client.add_app(app_def)
+
+    common.deployment_wait(service_id=app_id)
+
+    tasks = client.get_tasks(app_id)
+    assert len(tasks) == 1, "The number of tasks is {} after deployment, but 1 was expected".format(len(tasks))
+
+    port = tasks[0]['ports'][0]
+    host = tasks[0]['host']
+    cmd = "curl {}:{}/data/foo".format(host, port)
+
+    @retrying.retry(wait_fixed=1000, stop_max_attempt_number=30, retry_on_exception=common.ignore_exception)
+    def check_task(cmd, target_data):
+        run, data = shakedown.run_command_on_master(cmd)
+
+        assert run, "{} did not succeed".format(cmd)
+        assert data == target_data, "'{}' was not equal to {}".format(data, target_data)
+
+    check_task(cmd, target_data='hello\n')
+
+    shakedown.kill_process_on_host(host, '[h]ttp.server')
+    common.deployment_wait(service_id=app_id)
```

Comment:
I'm not sure the `TASK_FAILED` triggers a new deployment. Also, it is a race condition. The `deployment_wait` can succeed before the deployment even started. How about using the event stream instead? @zen-dog was really enthusiastic about this ;).

Reply:
Indeed, killing a task/instance does not trigger a deployment.

Reply:
And this is a hacky way to kill tasks(instances in pod test): we could achieve the same with `DELETE v2/apps/{app_id}/tasks/{task_id}` and `DELETE /v2/pods/{id}::instances/{instance}` and there are appropriate `client` methods for it. However even that way there is still no deployment triggered.

Reply:
Yeah, you are right, guys, it doesn't. I will address it. On the other hand, killing a task using the Marathon API is not the same a killing a task process. So, it is not a hacky way. It is just what needs to be done.

Reply:
How do you mean? Marathon will send a task kill request to mesos so what's the difference? 

The problem with current approach: you kill *all* `http.server`s on the agent *(actually you kill task's executor and then fail to kill the task itself since it's gone already)* which means this test can't run parallel to any other test. Not that we're doing this right now but a man can dream.

Reply:
What I mean is that if there is a pod of two tasks, and if Marathon asks Mesos to kill one of them, both of the tasks will transition to `TASK_KILLING` state, and then `TASK_KILLED`. On the other hand, if one of the tasks gets killed not by Mesos or just fails, it will transition to `TASK_FAILED`, and the other one will be killed by Mesos, and therefore the state transition for the other task will be `TASK_KILLING -> TASK_KILLED`. I understand the flaws of the current approach, but this is what I would like to have tested.

Reply:
I understand the reasoning but still don't like the idea of killing all `http.server` on that machine. how about putting some uuid into the cmd and grepping for it? e.g. smth. like `f"echo {uuid4().hex} && python http.server..."`?